### PR TITLE
Update generate password logic to use proper methods

### LINKED ISSUE #15740: Use strong random number/string generators inside generatePassword()
### Summary

Looking for a  native solution to create random strings, I stumbled upon the generate Password function today.

I've noticed that the "randomness" of the generated Password relies on mt_rand(), a "Pseudo-Random Number Generator". I guess this is a pice of code that has been active from the old php5 days. PRNG are considered to be a "weak" nowadays.

Basing a random number on a timestamp using srand() is also considered to be a potential problem.

```
    source: core/model/modx/moduser.class.php, line 797
    /**
     * Returns a randomly generated password
     *
     * @param integer $length The length of the password
     * @param array $options
     * @return string The newly generated password
     */
    public function generatePassword($length = null,array $options = array()) {
        if ($length === null) {
            $length = $this->xpdo->getOption('password_generated_length', null, 10, true);
        }
        $passwordMinimumLength = $this->xpdo->getOption('password_min_length', null, 8, true);
        if ($length < $passwordMinimumLength) {
            $length = $passwordMinimumLength;
        }
        $options = array_merge(array(
            'allowable_characters' => 'abcdefghjkmnpqrstuvxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789',
            'srand_seed_multiplier' => 1000000,
        ),$options);

        $ps_len = strlen($options['allowable_characters']);
        srand((double) microtime() * $options['srand_seed_multiplier']);
        $pass = '';
        for ($i = 0; $i < $length; $i++) {
            $pass .= $options['allowable_characters'][mt_rand(0, $ps_len -1)];
        }
        return $pass;
    }
```

### Why is it needed?

Random passwords should be random by todays standards. 

I'm aware that its very unlikely that the login mechanism might be compromised by this "vulnerabilty", but I like to point out that a modx developer might make use of this function to generate a "secure" and "random" strings for other purposes.

### Suggested solution(s)

In php 7 "Cryptographically Secure Pseudo-Random  NUMBER/STRING Generators" are available. These methods are considered to be "strong":

* https://www.php.net/manual/de/function.random-bytes.php (php >= 7.0)
* https://www.php.net/manual/en/function.random-int.php (php >= 7.0)
* https://www.php.net/manual/de/function.openssl-random-pseudo-bytes.php (php >= 5.3.0)

PHP 5.x polyfill for random_bytes() and random_int() 
* https://github.com/paragonie/random_compat

SIDENOTE:
> The random_bytes() function returns a binary string which may contain the \0 character. This can cause trouble in several common scenarios, such as storing this value in a database or including it as part of the URL. The solution is to hash the value returned by random_bytes() with a hashing function such as md5 or sha1.
SOURCE: https://symfony.com/doc/current/components/security/secure_tools.html

### Discussion

* How important are random strings/numbers?
* Should modx provide them, or should developers create their own methods?
* Are there more random number/string generators inside the core that should be reviewed?

Cheers patrick

### PR DESCRIPTION
**Pull request by Mark-H**
### What does it do?

Re-up of #15894

Changes password generation method to be more secure.

### Why is it needed?
Actually random generation. 

### How to test
Apply and see passwords still get generated.

### Related issue(s)/PR(s)
Re-up of #15894
Fixes #15740

### CODE REVIEW

**Comment by smg6511** [comment]
I've got a couple general questions on this before doing a quick review:

- I assume that since you changed the password dictionary variable from `$options['allowable_characters']` to `$options['alphabet']` that you're not worried about breaking usage of this method by custom implementations/extras. If that's the case, we might want to consider deprecating the `$options` param or go even further an deprecate both params and consider this a private method.
- Re the following block, it seems to me you'd want to define the alphabet if the `$options['alphabet']` param _is_ empty; or am I not seeing this right:
```php
    if (!empty($options['alphabet'])) {
        $alphabet = array_merge(range('a', 'z'), range('A', 'Z'));
        shuffle($alphabet);
        return substr(implode($alphabet), 0, $length);
    }
```
in other words, is something like this what you're really going for?
```php
    $alphabet = empty($options['alphabet']) 
        ? array_merge(range('a', 'z'), range('A', 'Z')) 
        : $options['alphabet']
        ;
    shuffle($alphabet);
    return substr(implode($alphabet), 0, $length);
```

**Comment by Mark-H** [comment]
Hmm, I overlooked the change from `allowable_characters` to `alphabet` recreating the original PR, but the behavior seems to have been intended differently. I don't know if the old behavior is worth keeping or not. Would welcome feedback on this, as well as examples that may use this code a certain way that would break. 

I'd be fine with removing that block entirely and limiting the functionality of this method to only the new generation with `random_bytes()` to keep things simple.

**Comment by smg6511** [comment]
Yeah, my guess is there's likely been no usage of that method outside of the core. If, however, there has been and a dictionary were passed in via `$options['allowable_characters']`, that would break with the new var name. That said, I'd opt for dropping the block as you suggested.

**Comment by smg6511** [comment]
Looking at this again I see what you mean when you imply the change (changing the options key in question) wouldn't break anything. It might cause unexpected output if there was external use of this method where chars and/or a custom seed were passed into the $options, but would not prevent a password from being generated. I get now that your changed variable is meant to be a _boolean_ value, not a character set (in which case a key of something like 'use_alpha_chars' would be more clear). 

At any rate, I'd still say drop that block altogether because the new generation logic you added below it will create a better password anyway. Further, I'd even move toward making this method private and removing its params, making it even more clear that this is a utility method meant for internal use only. After all, I think the intention is that these passwords are temporary ones that users should immediately change.

**Comment by theboxer** [comment]
I'd agree with @smg6511 and nuke everything but the length & bin2hex usage :)

However I'd keep the signature same (even though `$options` will not be used), just to keep the BC going.

**Comment by theboxer** [comment]
@Mark-H @smg6511 I adjusted the PR to keep the `allowable_characters` option and use the random bytes for generating the password.

**Review by opengeek — APPROVED** [review]
